hasktorch/test/PipelineSpec.hs

```diff
@@ -0,0 +1,61 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleContexts #-}
+module PipelineSpec where
+
+import Control.Concurrent
+import Control.Concurrent.Async
+import Control.Monad
+import Pipes
+import Pipes.Concurrent
+import System.Exit
+import System.IO
+import System.Timeout
+import Torch.Data.Pipeline
+  
+import Test.Hspec
+
+
+defaultTimeout :: Int
+defaultTimeout = 100000  
+
+newtype MockData = MockData Int
+instance Dataset MockData Int where
+  getBatch (MockData iters) _ = pure $ iters
+  numIters (MockData iters)  = iters
+
+instance ConcurrentDataset MockData Int where
+  getBatchConcurrently _ = getBatch 
+
+testFoldTimeout :: MockData -> IO ()
+testFoldTimeout dataset = do
+  (toBatches, fromBatches, sealBatch) <- spawn' (bounded 1)
+  thread <- async $ runEffect $ readBatches dataset toBatches
+  fold <- pure $ foldFromProducer (takeBatch fromBatches)
+  timedOut <- async $ fold $ FoldM takeBatchThenTimeout (pure 0) pure
+  wait thread
+  cancel timedOut
+
+
+testConcurrentFoldTimeout :: MockData  -> Int -> IO ()
+testConcurrentFoldTimeout dataset numWorkers = do
+  (toBatches, fromBatches, sealBatch) <- spawn' (bounded numWorkers)
+  threads <- forM [1..numWorkers] $ \workerId -> async $ runEffect $ readBatchesConcurrently workerId dataset toBatches
+  fold <- pure $ foldFromProducer (takeBatch fromBatches)
+  timedOut <- async $ fold $  FoldM takeBatchThenTimeout (pure 0) pure
+  mapM_ wait threads
+  cancel timedOut
+
+takeBatchThenTimeout :: Int -> Int -> IO Int
+takeBatchThenTimeout _ input = threadDelay 50000 >> pure input
+
+runTest :: IO () -> IO (Maybe ())
```

Comment:
can you add a little doc string that explains that `Nothing` corresponds to a timeout, please?